module-apps/apps-common/ApplicationCommon.cpp

```diff
@@ -772,243 +801,113 @@ namespace app
+            auto retval = request->handle();
+            if (not retval) {
+                LOG_ERROR("Popup %s handling failure, please check registered blueprint!", popup.c_str());
+            }
+            return retval;
+            LOG_ERROR("no blueprint for popup: %s", popup.c_str());
         }
+        return false;
     }
 
+    // oh this is kind of bad... we require popup start window to properly abort selected popup XD
     void ApplicationCommon::abortPopup(gui::popup::ID id)
     {
         const auto popupName = gui::popup::resolveWindowName(id);
-
-        if (getCurrentWindow()->getName() == popupName) {
-            returnToPreviousWindow();
-        }
-        else {
-            popWindow(popupName);
-        }
+        LOG_INFO("abort popup: %s from window %s", popupName.c_str(), getCurrentWindow()->getName().c_str());
+        windowsStack().popTo(popupName);
+        returnToPreviousWindow();
     }
 
-    bool ApplicationCommon::isPopupPermitted([[maybe_unused]] gui::popup::ID popupId) const
+    bool ApplicationCommon::handleUI_DBNotification(sys::Message *msg, const UiNotificationFilter &filter)
     {
-        return true;
-    }
-
-    bool ApplicationCommon::popToWindow(const std::string &window)
-    {
-        if (window == gui::name::window::no_window) {
-            bool ret = false;
-            if (windowsStack.stack.size() <= 1) {
-                windowsStack.stack.clear();
-                ret = true;
+        bool handled = false;
+        for (const auto &[name, window] : windowsStack()) {
+            if (filter == nullptr || (filter != nullptr && filter(msg, window->getName()))) {
+                handled |= window->onDatabaseMessage(msg);
             }
-            return ret;
-        }
-
-        auto ret = std::find(windowsStack.stack.begin(), windowsStack.stack.end(), window);
-        if (ret != windowsStack.stack.end()) {
-            LOG_INFO("Pop last window(s) [%d] :  %s",
-                     static_cast<int>(std::distance(ret, windowsStack.stack.end())),
-                     ret->c_str());
-            windowsStack.stack.erase(std::next(ret), windowsStack.stack.end());
-            LOG_INFO("Curent window... %s vs %s", ret->c_str(), windowsStack.stack.back().c_str());
-            return true;
         }
-        return false;
+        return handled;
     }
 
-    void ApplicationCommon::pushWindow(const std::string &newWindow)
+    /// TODO change how we handle "" actually ... how do we handle "" in the first place?
+    void ApplicationCommon::pushWindow(const std::string &newWindow, const gui::popup::Disposition &d)
     {
         // handle if window was already on
-        LOG_DEBUG("App: %s window %s request", GetName().c_str(), newWindow.c_str());
-        if (popToWindow(newWindow)) {
+        LOG_INFO("App: %s window %s request", GetName().c_str(), newWindow.c_str());
+        if (newWindow.empty() && windowsStack().popLastWindow()) {
```

Comment:
why we try to clear stack here on empty request?

Reply:
Probably popLastWindow is wrong name ?

Reply:
@PrzeBrudny this is STL, I'm afraid rewriting names in the standard library is out of my reach. 
![image](https://user-images.githubusercontent.com/6411862/144007173-bb72d60a-e70d-4ac9-a738-3105cd712093.png) of course if you are able to convince others, please feel free to propose a better naming convention to the committee, i.e. the way it was done[ here ](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1754r0.pdf):) 

popLastWindow does exactly what the name implies. perform pop on the last window. On success, -> returns true, on fail return false. The best would be to not have it at all in `ApplicationCommon` but it would be too much rework.